add -a flag. Task 2 & 4

abdul-moize opened the pull request:
add -a flag to weatherman. this flag calculates averages of extreme temperatures(highest lowest) and average  mean humidity for a given month. also implemented task 4. which means user can execute multiple flags at once

to test task2 follow these steps:
1. pull this branch in your local repo
2. run this command: `weatherman.py -a 2006/6`
you can change directories path by doing this: `weatherman.py path/to/files -a 2006/6`

to test task4 follow these steps:
1. pull this branch
2. run this command: 'weatherman.py -a 2005/6 -e 2005'
PS: you can nest as many flags as you want the program will generate the output on the screen

you can also specify date like this `2007/07` whichever you are more comfortable with
please try different values of date and flag an if there are any issues do let me know

abdul-moize committed:
add -a flag

abdul-moize committed:
resolve conflicts

abdul-moize committed:
requested change
add None returns and update accordingly

abdul-moize committed:
fix pylint errors

abdul-moize committed:
fix docstring

abdul-moize committed:
fix docstring

abdul-moize committed:
fix docstring

abdul-moize committed:
add default path

arslanashraf7 commented on .gitignore:
```diff
@@ -1,2 +1,4 @@
\ No newline at end of file
+.idea/
+__pycache__/
+modules/__pycache__
```
Make sure we add one new line at the end of the file. This will apply to all the other similar places too.

abdul-moize replied:
done

arslanashraf7 commented on modules/average_month.py:
```diff
@@ -0,0 +1,109 @@
+"""
+This module generates a report which displays average highest and lowest temperature
+as well as average max humidity for a given month
+"""
```
Can you confirm from the task details if we need to show avg max humidity or mean humidity?

```suggestion
"""
This module generates a report for the average highest and lowest temperature
as well as average mean humidity for a given month
"""
```

abdul-moize replied:
its mean humidity. fixed it

arslanashraf7 commented on modules/average_month.py:
```diff
@@ -0,0 +1,109 @@
+
+def calculate_averages(pattern, path):
+    """
+    This function reads all the files at the given path which contain the given pattern
+    from each file that matches the pattern finds the following:
```
The function that reads the files is `read_data`, this function should contain what it does and not what's happening in `read_data()` which is being called. Can you try to update this doc string?

abdul-moize replied:
updated

arslanashraf7 commented on modules/average_month.py:
```diff
@@ -0,0 +1,109 @@
+    sum_mean_humidity = 0
+    lines = []
+    for lines in read_data(pattern, path):
+        if lines is None:
+            return None
+        for line in lines:
```
We can just continue the logic if lines exist and at the end of the function, we can just return None if nothing was returned from if above.

```suggestion
        if lines:
            for line in lines:
```

abdul-moize replied:
done

arslanashraf7 commented on modules/average_month.py:
```diff
@@ -0,0 +1,109 @@
+                sum_lowest_temperature += min_temperature
+            mean_humidity = get_mean_humidity(parsed_line)
+            if mean_humidity is not None:
+                sum_mean_humidity += mean_humidity
```
Can you try this? You might not need to use a None check above it. This would apply to all the other places as well.

```suggestion
                sum_mean_humidity += mean_humidity or 0
```

abdul-moize replied:
this is great. applied to all 3 sums

arslanashraf7 commented on modules/average_month.py:
```diff
@@ -0,0 +1,109 @@
+    Calculates and displays the average highest, lowest temperatures
+    and average mean humidity for a given month in given path
+    Args:
+        month(str): a value containing 4 digit year and 2 digit month like: '2002/02', '2003/04'.
```
I think you copied over and forgot the Updates?

abdul-moize replied:
i don't understand this comment. kindly elaborate

arslanashraf7 replied:
Is the year reference correct here? we should name this var as a month if it also contains the year.
`        month(str): a value containing 4 digit year and 2 digit month like: '2002/02', '2003/04'.
`

abdul-moize replied:
did u mean should not? i have renamed the variable to year_month to make it more clear

arslanashraf7 commented on modules/average_month.py:
```diff
@@ -0,0 +1,109 @@
+
+def get_averages_month(month, path=constants.WEATHER_FILES_DIR):
+    """
+    Calculates and displays the average highest, lowest temperatures
```
Refer to a comment i added above.

abdul-moize replied:
changed the function name and argument name

arslanashraf7 commented on modules/average_month.py:
```diff
@@ -0,0 +1,109 @@
+    and average mean humidity for a given month in given path
+    Args:
+        month(str): a value containing 4 digit year and 2 digit month like: '2002/02', '2003/04'.
+        path(str): a value containing path like: 'weatherfiles/'
```
```suggestion
        path(str): a value containing path for weather files e.g. : 'weatherfiles/'
```

abdul-moize replied:
done

arslanashraf7 commented on modules/average_month.py:
```diff
@@ -0,0 +1,109 @@
+                        Or
+                        None for failure
+    """
+    months_list = [
```
Can we do something to reuse the months_list here and https://github.com/abdul-moize/weather-man/blob/master/modules/extreme_year.py#L149. Maybe we could take benefit from dict/tuples ?. Also, can we move this to constants too?

abdul-moize replied:
i have moved the list to constants and used that list in both the files. i don't see any benefit by using dict/tuples. please elaborate on this

arslanashraf7 replied:
It was an option if we need to use multiple month name formats to support based on the index without using date-time formatted. Since you were using Full name months list at one place and 3 letter name on other.

arslanashraf7 commented on modules/utils.py:
```diff
@@ -33,3 +35,72 @@ def read_data(year, path):
+                            None if there is no entry or wrong entry
+    """
+    try:
+        return int(line[8])
```
Let's define these index values in constants. e.g. `MEAN_HUMIDITY_INDEX` and others similarly.

abdul-moize replied:
done

abdul-moize committed:
requested changes
fix docstrings
move months_list to contants
add indexes to contants
simplify sums in average_month.py

abdul-moize committed:
fix isort, black, pylint errors

abdul-moize committed:
renamed argument name

abdul-moize committed:
fix pylint error

abdul-moize committed:
rename function

arslanashraf7 commented on .isort.cfg:
```diff
@@ -0,0 +1,2 @@
+[settings]
+profile = black
```
New line at the end of the file?

abdul-moize replied:
sorry. added now

arslanashraf7 commented on modules/average_month.py:
```diff
@@ -0,0 +1,93 @@
+    2. avg lowest temperature
+    3. avg mean humidity
+    Args:
+        pattern(str): a string which contains a year/month e.g '2006/5', '2007/3'
```
Let's make it even cleaner, Pass year and month as 2 separate params. Also, it would be a good call to check the validation on month and year e.g. either they were parsed or not -- This validation could be done in the caller function or through a generic utils method.

arslanashraf7 commented on modules/average_month.py:
```diff
@@ -0,0 +1,93 @@
+    sum_mean_humidity = 0
+    lines = []
+    for lines in read_data(pattern, path):
+        if lines:
```
If lines is None would the loop continue? Do we need this check here?

abdul-moize replied:
No, the loop would not continue. we don't need this check here. removed it

arslanashraf7 commented on weatherman.py:
```diff
@@ -67,16 +61,14 @@ def main():
             iteration = 0
             continue
         iteration += 1
+        print()
```
Is it doing anything specific?

abdul-moize replied:
it just add a new line to make it easier to differentiate between each report

abdul-moize committed:
add newlines at the end

abdul-moize committed:
update regex, change arguments of calculate_averages()

abdul-moize committed:
update regex, add argument checks, move constants to constants.py

abdul-moize committed:
add validators.py, update weatherman.py, update extreme_year.py, update average_month.py
